File: java/testsuite/Jogl4Examples/MeshExample/src/gui/TangibleInterfaceInteractionTechniques.java
package gui;

import vsdk.toolkit.gui.tangibleInterfaces.TangibleInterfaceEvent;
import vsdk.toolkit.gui.tangibleInterfaces.TangibleInterfaceEvent2RayGizmoMapper;
import vsdk.toolkit.gui.tangibleInterfaces.TangibleInterfaceListener;
import model.MeshModel;

public class TangibleInterfaceInteractionTechniques implements TangibleInterfaceListener {
    private static final String RAY_CUBE_TANGIBLE_ELEMENT_ID = "rayCube";

    private final MeshModel model;
    private final Runnable repaintCallback;
    private final TangibleInterfaceEvent2RayGizmoMapper mapper;

    public TangibleInterfaceInteractionTechniques(MeshModel model, Runnable repaintCallback) {
        this.model = model;
        this.repaintCallback = repaintCallback;
        this.mapper = new TangibleInterfaceEvent2RayGizmoMapper(model.getCamera());
    }

    @Override
    public void tangibleInterfaceEventReceived(TangibleInterfaceEvent event) {
        if ( event == null ) {
            return;
        }

        if ( !RAY_CUBE_TANGIBLE_ELEMENT_ID.equals(event.getId()) ) {
            return;
        }

        mapper.map(event, model.getRayGizmo());
        repaintCallback.run();
    }
}

File: cpp/testsuite/OpenGL4Examples/MeshExample/src/gui/TangibleInterfaceInteractionTechniques.cpp
#include "gui/TangibleInterfaceInteractionTechniques.h"
#include "model/MeshModel.h"
#include "vsdk/toolkit/gui/tangibleInterfaces/TangibleInterfaceEvent.h"

const char* TangibleInterfaceInteractionTechniques::RAY_CUBE_TANGIBLE_ELEMENT_ID = "rayCube1";

TangibleInterfaceInteractionTechniques::TangibleInterfaceInteractionTechniques(
    MeshModel* model, std::function<void()> repaintCallback)
    : model(model),
      repaintCallback(repaintCallback),
      mapper(model->getCamera())
{
}

void TangibleInterfaceInteractionTechniques::tangibleInterfaceEventReceived(const TangibleInterfaceEvent& event)
{
    if ( event.getId() != RAY_CUBE_TANGIBLE_ELEMENT_ID ) {
        return;
    }

    mapper.map(event, model->getRayGizmo());
    repaintCallback();
}
